neural/run_workflow.py

```diff
@@ -614,3 +631,158 @@ def test_run(q_docpartitions, q_fold_config_map, bertmodel, train_val_dir, test_
+    docid_attnweights_map = {dsettype: {} for dsettype in data_loaders if dsettype in {'validation', 'test'}}
+    # store sentences' attention weights
+
+    if(os.path.isfile(os.path.join(sents_embed_dir, 'bert_proc_docs.pkl'))):
+        bert_proc_docs = ReaderWriter.read_data(os.path.join(sents_embed_dir, 'bert_proc_docs.pkl'))
+    else:
+        bert_proc_docs = {}
+
+    print("device: {} | question: {} | fold_num: {} | dsettype: {} | pid: {}"
+          "".format(device, options.get('question'), fold_num, dsettype, pid))
+    pred_class = []
+
+    data_loader = data_loaders[dsettype]
+
+    for m, m_name in models:
+        m.eval()
+
+    sample_counter = 0
+    for i_batch, samples_batch in enumerate(data_loader):
+
+        logprob_scores = []
+
+        docs_batch, docs_len, docs_sents_len, docs_attn_mask, docs_labels, docs_id = samples_batch
+
+        docs_batch = docs_batch.to(device)
+        docs_attn_mask = docs_attn_mask.to(device)
+        docs_sents_len = docs_sents_len.type(torch.int64).numpy()  # to feed this in RNN
+        docs_labels = docs_labels.type(torch.int64).to(device)
+
+        with torch.set_grad_enabled(dsettype == 'train'):
+            # print("number of examples in batch:", docs_batch.size(0))
+            num_docs_perbatch = docs_batch.size(0)
+            for doc_indx in range(num_docs_perbatch):
+                # print('doc_indx', doc_indx)
+                doc_id = docs_id[doc_indx].item()
+                if(doc_id in bert_proc_docs):
+                    # due to GPU limit
+                    embed_sents = torch.load(bert_proc_docs[doc_id], map_location=device)
+                else:
+                    embed_sents = bert_encoder(docs_batch[doc_indx], docs_attn_mask[doc_indx],
+                                               docs_len[doc_indx].item())
+                    # add embedding to dict
+                    embed_fpath = os.path.join(sents_embed_dir, '{}.pkl'.format(doc_id))
+                    ReaderWriter.dump_data(embed_sents, embed_fpath)
+                    bert_proc_docs[doc_id] = embed_fpath
+
+                sents_rnn_hidden = sent_encoder(embed_sents, docs_sents_len[doc_indx],
+                                                docs_len[doc_indx].item())
+
+                # # remove the embedding from GPU
+                # bert_proc_docs[doc_id].to(cpu_device)
+                # print('sents_rnn_hidden', sents_rnn_hidden.shape)
+                enc_sents = sents_rnn_hidden
+                # print('enc_sents', enc_sents.shape)
+                doc_out, doc_attn_weights = doc_encoder(enc_sents)
+                # print('doc_out', doc_out.shape)
+                # print('doc_attn_weights', doc_attn_weights.shape)
+                # tracking attention weight for validation and test examples
+                if(dsettype in docid_attnweights_map):
+                    docid_attnweights_map[dsettype][doc_id] = doc_attn_weights
```

Comment:
I am not sure if it is the place to discuss this -- we discussed before to keep show/keep attention when the prediction of the model is `1` (i.e. the criterion is fulfilled). One option is to not process the attention weights here and do the processing when we try to output the prediction (i.e. in case it is `1` call `return_attnw_over_sents` or another dedicated function)

Reply:
I agree we should only show attended sentences when prediction = 1. However, I would like to leave the decision to show the attention results to whatever is consuming the prediction (such as the website itself). That gives us more flexibility. For example, we may decide in the future to only show predictions of 1 and their attended sentences on the website when the prediction's logprob score meets a threshold (only if the model is "confident"). By always returning the attention weights as above, making such a change would only involve changing high level website code, not this low level model code.